File: src/main/java/com/rdvdev2/timetravelmod/api/dimension/TimelineBuilder.java
package com.rdvdev2.timetravelmod.api.dimension;

import com.rdvdev2.timetravelmod.impl.common.dimension.timeline.Timeline;
import net.minecraft.item.Item;
import net.minecraft.util.registry.RegistryKey;
import net.minecraft.world.World;

public class TimelineBuilder {

    private int minTier = 1;
    private RegistryKey<World> world = null;
    private Item icon = null;

    public TimelineBuilder setMinTier(int minTier) {
        this.minTier = minTier;
        return this;
    }

    public TimelineBuilder setWorld(RegistryKey<World> world) {
        this.world = world;
        return this;
    }

    public TimelineBuilder setIcon(Item icon) {
        this.icon = icon;
        return this;
    }

    public ITimeline build() {
        if (world == null || icon == null)
            throw new RuntimeException("You must define all properties");
        return new Timeline(minTier, world, icon);
    }
}

File: src/main/kotlin/com/rdvdev2/timetravelmod/api/dimension/TimelineBuilder.kt
package com.rdvdev2.timetravelmod.api.dimension

import com.rdvdev2.timetravelmod.impl.common.dimension.timeline.Timeline
import net.minecraft.item.Item
import net.minecraft.util.registry.RegistryKey
import net.minecraft.world.World

class TimelineBuilder {

    private var minTier = 1
    private var world: RegistryKey<World>? = null
    private var icon: Item? = null

    fun setMinTier(minTier: Int): TimelineBuilder {
        this.minTier = minTier
        return this
    }

    fun setWorld(world: RegistryKey<World>?): TimelineBuilder {
        this.world = world
        return this
    }

    fun setIcon(icon: Item?): TimelineBuilder {
        this.icon = icon
        return this
    }

    fun build(): ITimeline {
        if (world == null || icon == null) throw RuntimeException("You must define all properties")
        return Timeline(minTier, world!!, icon!!)
    }
}